managed-pointer/cpp-20/D0567R1.md

```diff
@@ -5,222 +5,384 @@
+### 2.6. Must allow optimisation of data movement between devices
 
-Any ``managed_ptr`` can only have a single host CPU that can be accessible for any given application, that is capable of triggering synchronisation operations.
+A unified interface for data movement must provide interoperability bewteen implementations, however this will likely be inneficient as it will likely require a middle man stage that facilitates the required operations of each implementation.
 
-*Note: some systems may wish to have multiple host CPU nodes which are capable of triggering synchronisation operations. This feature is discussed in more detail in the future work section.*
+Therefore it's important that a unified interface for data movement provide a facility for customising data movement between particular memory regions in order to optimise those operations rather than relying on a generic middle man.
 
-If the host CPU is not accessible the ``managed_ptr`` is required to maintain a pointer to the accessible **execution context** in order for it to perform synchronisation operations.
+### 2.7. Must abstract device capabilities
 
-Memory is only required to be allocated where it is accessible. The managed memory allocation is only required to be allocated on the host CPU memory region if the ``managed_ptr`` is constructed with a pointer or if a synchronisation operation is triggered to the host CPU. Managed memory allocation is only required to be allocated on the memory region of an **execution context** if a synchronisation operation is triggered to that **execution context**.
+Many heterogeneous architectures such as GPGPUs, DSPs or FPGAs have considerable restrictions on the code that is executed on the compared to that of a CPU, such no capability to dynamically allocate memory or signal directly with other devices during execution and the level of forward proress guarantees the device can provide. This means that a unified inteface for data movement needs to take into consideration these restrictions in the design of how memory on devices is allocated and synchronised.
+
+A typical example of this is that a CPU cluster would have no problem allocating a `managed_ptr` on any ne of it's nodes and peforming synchronisation operations on any node, where as with a typical GPU all allocation and sycnhronisation must be done from a host CPU device priory to execution.
+
+## 3. Proposed Additions
+
+### 3.1. Header `<execution>` synopis
 
 ```cpp
 namespace std {
 namespace experimental {
+inline namespace concurrency_v2 {
 namespace execution {
 
+/* extensions to executor classes */
+class executor {
+  ...
+
+  /* aliases */
+  using global_memory_allocator_type = __global_memory_allocator_type__;
```

Comment:
Tbh I wasn't sure about this name, though I couldn't think of a better term to describe it. This is the allocator for allocating memory within the executor's memory region. Perhaps `executor_memory_allocator_type` would be better.